pkg/instrumentation/database.go

```diff
@@ -10,20 +10,26 @@ import (
 	"gorm.io/gorm"
 )
 
-const (
+type spanContextKey string
+
+var (
 	// ParentSpanGormKey is the name of the parent span key
-	ParentSpanGormKey = "tracingParentSpan"
+	parentSpanGormKey = spanContextKey("trancingParentSpan")
 	// SpanGormKey is the name of the span key
-	SpanGormKey = "tracingSpan"
+	spanGormKey = spanContextKey("tracingSpan")
 )
 
 // TraceDatabase sets span to gorm settings, returns cloned DB
 func TraceDatabase(ctx context.Context, db *gorm.DB) *gorm.DB {
 	if ctx == nil {
 		return db
 	}
+
 	parentSpan, _ := ddtrace.SpanFromContext(ctx)
-	return db.Set(ParentSpanGormKey, parentSpan)
+
+	return db.Session(&gorm.Session{
+		Context: context.WithValue(db.Statement.Context, parentSpanGormKey, parentSpan),
```

Comment:
[q]: why can't keep using the [Set](https://pkg.go.dev/gorm.io/gorm#DB.Set) function here? Looks like the [dd-trace-go wrapper](https://github.com/DataDog/dd-trace-go/blob/48141787da9e038992e67c66768b722f1c7713c8/contrib/jinzhu/gorm/gorm.go#L82) uses the `Set/Get` everywhere for the `db` context.

Reply:
That's the `jinzhu` one you're mentioning, the one we use now is `gorm.io/gorm` following [same pattern](https://github.com/scribd/go-sdk/pull/69).
On why it doesn't work, using `db.Set()` put data into `db.Statement.Settings` which doesn't get copied while copying in `db.Statement.clone()` [here](https://github.com/go-gorm/gorm/blob/da2b2861de47900edc6d0b1898bbdd5d5381b412/gorm.go#L242).

Reply:
@laynax 
>That's the jinzhu one you're mentioning, the one we use now is gorm.io/gorm following

Oh right, thanks for pointing this out 👍